src/manifold/deferred.clj

```diff
@@ -1386,6 +1512,152 @@
 
 (prefer-method print-method IDeferred IDeref)
 
+(defn- flatten-deferred
+  "Converts a Deferred<Deferred<T>> into a Deferred<T>"
+  [value]
+  (chain value identity))
+
+(defn- async-for
+  "Retuns a CompletionStage async version of the given operator"
+  [original]
+  (fn result
+    ([this operator]
+     (result this operator (or (ex/executor) (ex/execute-pool))))
+    ([this operator executor]
+     (flatten-deferred
+      (future-with executor (original this operator))))))
+
+(defn- async-for-dual
+  "Retuns a CompletionStage async version for the given two deferred operator"
+  [original]
+  (fn result
+    ([this other operator]
+     (result this other operator (or (ex/executor) (ex/execute-pool))))
+    ([this other operator executor]
+     (flatten-deferred
+      (future-with executor (original this other operator))))))
+
+(defmacro ^:no-doc assert-some
+  "Throws NullPointerException if any of the aruments is null."
+  [& values]
+  `(do ~@(for [value values]
+         `(when (nil? ~value)
+            (throw (NullPointerException. ~(str value " was null"))))
+         )))
+
+;; we do success-deferred on the result so that
+;; if the original operator returns a deferred, chain
+;; won't automatically flatten it out.
+(defn- then-apply [this ^Function operator]
+  (assert-some operator)
+  (chain this #(success-deferred
+                (.apply operator %))))
```

Comment:
Why are you using success-deferred here, but not elsewhere?

I'm increasingly wondering if `chain` is right for the job. It may be better to use some mixture of `on-realized`, `success-error-unrealized`, `unwrap`, etc.

`chain` is designed for multiple fns, and it's designed to let users utilize non-manifold fns to proess deferreds without having to think about them much. But if it's unwrapping behavior is causing problems, then it's not the right tool here.

Reply:
> Can you explain your use of success-deferred and flatten-deferred? I'm trying to understand why you're unwrapping or rewrapping return values.

- About `chain` and `success-deferred`

`chain` behaves both like `fmap/map` and `bind/mapcat` for manifold deferreds depending on the return of the given function. This means that for `thenApply`, if the given `Function` ends up returning another deferred, doing something like `(chain d #(.apply f %))` would be problematic since this `thenApply` call would behave like a `.thenCompose` call. If we use `success-deferred` on the result, `chain` will behave like `bind/mapcat`, and it will flatten out the `SuccessDeferred<OriginalValue>` into `OriginalValue` for the resulting deferred, making it consistently behave like `map/fmap`. This is exactly like using bind/`>>=` in haskell with `return` to get `fmap` or `liftM`. This is not a problem in most other places since the return value for those is void/nil.

- About flatten-deferred

```clojure
(flatten-deferred
      (future-with executor (original this operator)))
```
In this call, `(original this operator)` returns a `Deferred<T>` and `future-with` returns a `Deferred<Deferred<T>>`
so we have to flatten it out to a `Deferred<T>` to get the expected results. It's less about `chain` and more about `future-with`.

> I'm increasingly wondering if chain is right for the job. It may be better to use some mixture of on-realized, success-error-unrealized, unwrap, etc.

I tried using solely the public API functions and macros for the implementation of these. Maybe we should implement a map-only `chain` for `thenApply` and possibly expose it as a public API function as well?

Reply:
Unfortunately, I have yet to play around with Haskell, so I'm not familiar with your example. One day!

Since you're messing with the guts of Manifold, definitely **don't** restrict yourself to the public API. 

Also, let's align on some terminology. `flatten` is a Clojure core fn that turns a tree into a sequence (e.g., `[1 [2 [3 4]]] => (1 2 3 4)`). What we're doing here is called "unwrapping" in Manifold in general. The `unwrap/unwrap'` fns recursively unbox nested+realized deferreds until they reach either a non-deferred value, or an unrealized deferred. If it's just a single level of unwrapping, then it could also be called derefing. 

I'm going to rename the params in the examples, because they're confusing my Clojure-trained brain 😄 . (Unfortunately, I don't have a great name repalcement for `operator`; I'd call it `f` for a function, but sometimes it's a stupid Java verb-noun like Consumer or Runnable. Ugh.)

----

One solution is really simple. Instead of derefing the `future-with` deferred, deref _inside_ the `future-with`, like so:

```clojure
;; Replace
(flatten-deferred
      (future-with executor (orig-fn d operator)))

;; with
(future-with executor @(orig-fn d operator))
```

This will work just fine, since we can assume the orig-fns always return a deferred.

----

There's another possibility. 

The various CompletionStage `*Async` methods are flexible in _how_ they handle asynchrony, just as long as they _do_. One way we could do this is, instead of using code like

```clojure
(flatten-deferred
  (future-with executor (original this operator)))
```
to turn synchronous versions async, instead we could use `onto` and `on-realized` like

```clojure
(-> this 
    (onto executor) 
    (on-realized (fn [val] (original val operator)) 
                 (fn [_]))))  ;; empty fn is in case we're doing nothing for the error/exception path
```

We'd then change all the helper fns to take an optional executor, like so:

```clojure
(defn- then-accept
  ([d ^Consumer c]
   (then-accept d c nil))

  ([d ^Consumer c executor]
   (assert-some c)
   (let [d (if executor
             (onto d executor)
             d)
         d' (deferred)]
     (on-realized d
                  (fn [val]
                    (success! d' (.accept c val)))
                  (fn [_]))
     d')))
```

Now we have a `then-accept` function that works for _both_ sync and async cases. 

Given there's still a lot of commonality between the various helper methods, we'd want to factor out common functionality, or maybe use macros, paramaterizing by the actual method (like `.accept()`), and probably normal/exceptional/both. (Maybe more.)

This would fit how a lot of the rest of Manifold works. `onto` creates a copy of the deferred that guarantees its callbacks will be run on the executor, `on-realized` sets up those callbacks, and `d'` is created to return the eventual value.

It's a bit wordier, I admit, but you're no longer fighting against `chain`, which is not designed for this.

Reply:
Hmm, I thought of one problem in your original code. `chain` is going to unwrap *all* the nested deferreds (and return a new deferred with the final result), but CompletionStage methods wouldn't do that. CompletionStage works with only a single level of wrapping at a time, so if they were given a double-nested deferred, chain will provide them the wrong value.

Reply:
Oh, I didn't know `chain` unwrapped all level, my bad. Is it ok if I include a documentation change to `chain` about this detail as well?

Reply:
Yeah, it's documented in there, but kind of obliquely; the language could definitely be improved. 

The assumption of `chain/chain'` is that the fns you're passing to it are fns with non-deferred arguments, though they may _return_ deferreds. `chain` fully unwraps any deferred returned before handing it off to the next fn, and if it's not realized yet, it "pauses" the chain until the deferred is realized (it "pauses" by creating an `on-realized` callback on that deferred with the remainder of the chain, so it'll resume later).